src/utils/all_reduce.hpp

```diff
@@ -0,0 +1,106 @@
+//========================================================================================
+// (C) (or copyright) 2021. Triad National Security, LLC. All rights reserved.
+//
+// This program was produced under U.S. Government contract 89233218CNA000001 for Los
+// Alamos National Laboratory (LANL), which is operated by Triad National Security, LLC
+// for the U.S. Department of Energy/National Nuclear Security Administration. All rights
+// in the program are reserved by Triad National Security, LLC, and the U.S. Department
+// of Energy/National Nuclear Security Administration. The Government is granted for
+// itself and others acting on its behalf a nonexclusive, paid-up, irrevocable worldwide
+// license in this material to reproduce, prepare derivative works, distribute copies to
+// the public, perform publicly and display publicly, and to permit others to do so.
+//========================================================================================
+#ifndef UTILS_ALL_REDUCE_HPP_
+#define UTILS_ALL_REDUCE_HPP_
+
+#include <vector>
+
+#include <config.hpp>
+#include <parthenon_mpi.hpp>
+
+namespace parthenon {
+
+enum class Op { MAX, MIN, SUM, PROD, LAND, BAND, LOR, BOR, LXOR, BXOR, MAXLOC, MINLOC };
+#ifndef MPI_PARALLEL
+#define MPI_
+#endif
+
+template <typename T, int N = 0>
+struct AllReduce {
+  T val;
```

Comment:
I see that currently this only reduces on one value. Could this be extended to reduce on multiple values?

Reply:
I haven't tested it yet, but this should be able to support doing reductions on, for example, a `std::vector`.    See the last version of `StartReduce`.

Reply:
just pushed an update.  the vector thing is still untested, but perhaps the intent is clearer now.  would doing a reduction on a std::vector work for you?

Reply:
ok the poisson example now (needlessly) does a reduction on a vector of ints as an example.